src/log/CustomLog.cc

```diff
@@ -7,12 +7,167 @@
 #include "log/File.h"
+#include "Parsing.h"
+#include "sbuf/Stream.h"
+#include "SquidConfig.h"
+
+FormattedLog::~FormattedLog()
+{
+    close(); // TODO: destructing a Logfile object should be enough
+    aclDestroyAclList(&aclList);
+    safe_free(filename);
+    // leave logFormat alone -- we do not own that object
+}
 
 bool
-CustomLog::usesDaemon() const
+FormattedLog::usesDaemon() const
 {
     return (filename && strncmp(filename, "daemon:", 7) == 0);
 }
 
+void
+FormattedLog::parseOptions(ConfigParser &parser, const char *defaultFormatName)
+{
+    const char *explicitFormatName = nullptr;
+    char *key = nullptr;
+    char *value = nullptr;
+    while (parser.optionalKvPair(key, value)) {
+
+        if (strcmp(key, "on-error") == 0) {
+            if (strcmp(value, "die") == 0) {
+                fatal = true;
+            } else if (strcmp(value, "drop") == 0) {
+                fatal = false;
+            } else {
+                throw TextException(ToSBuf("unsupported ", cfg_directive, " on-error value: ", value,
+                                           Debug::Extra, "expected 'drop' or 'die'"), Here());
+            }
+            continue;
+        }
+
+        if (strcmp(key, "buffer-size") == 0) {
+            parseBytesOptionValue(&bufferSize, "bytes", value);
+            continue;
+        }
+
+        if (strcmp(key, "rotate") == 0) {
+            rotationsToKeep = Optional<unsigned int>(xatoui(value));
+            continue;
+        }
+
+        if (strcmp(key, "logformat") == 0 && defaultFormatName) {
+            if (explicitFormatName)
+                throw TextException(ToSBuf("duplicated ", cfg_directive, " option: ", key), Here());
+
+            explicitFormatName = value;
+            continue;
+        }
+
+        throw TextException(ToSBuf("unsupported ", cfg_directive, " option: ", key, "=", value), Here());
+    }
+
+    if (const auto formatName = explicitFormatName ? explicitFormatName : defaultFormatName) {
+        assert(defaultFormatName); // this log supports logformat=name
+        setLogformat(formatName);
+    } // else OK: this log does not support logformat=name and none was given
+}
+
+void
+FormattedLog::dumpOptions(std::ostream &os) const
+{
+    /* do not report defaults */
+
+    // TODO: Here and elsewhere, report both explicitly configured settings and
+    // various defaults. Properly excluding defaults requires wrapping most
+    // non-pointer members in Optional _and_ adding methods to compute the final
+    // option value after accounting for defaults (and those may change with
+    // reconfiguration!). And all that effort may still not result in a faithful
+    // reproduction of the original squid.conf because of size unit changes,
+    // order changes, duplicates removal, etc. More importantly, these reports
+    // are much more useful for determining complete Squid state (especially
+    // when triaging older Squids with some difficult-to-figure-out defaults).
```

Comment:
Much of the text here is relevant to admin using the report rather than dev coding. So maybe a new wiki page documenting what the 'mgr:config' report produces?
https://wiki.squid-cache.org/Features/CacheManager#Available_Reports

Reply:
> This is a general blog about config dumping. not specific to this object. It should not be here.

The problems that triggered (and that are mentioned in) this TODO are present in the source code this TODO is currently attached to (look for `if` statements below the comment). There ought to be other places with similar problems, but I cannot think of a _better_ place for this source code comment, and your first change request [version](https://github.com/squid-cache/squid/pull/824#discussion_r632888665) (quoted above) have not suggested any.

> Much of the text here is relevant to admin using the report rather than dev coding. So maybe a new wiki page documenting what the 'mgr:config' report produces?

At first, I thought that perhaps this change request was attached to the wrong source code comment -- the two did not match, but then I realized that you probably just missed the key word that the proposed comment started with: `TODO`. The proposed comment has no useful information for Squid administrators because it discusses how the _future_ code should work (and what problems it should solve), not how the existing code works today. Admins see what the existing code produces, of course.

If you insist on moving this TODO, please suggest a better place for it. If we cannot agree on a location (in Squid sources), then I will delete this comment despite my belief that it contains non-obvious[1] conclusions and would be quite useful to anyone (including myself) working on the documented problem in the future.

[1] Well, they may be obvious to everybody else, but their development took me some time.

Reply:
Hmm. I still dislike the TODO comment placement very much. But will accept this rational.